File: app/src/main/java/com/example/ctap/fido2/CredentialData.java
package com.example.ctap.fido2;

import com.fasterxml.jackson.dataformat.cbor.CBORFactory;
import com.fasterxml.jackson.dataformat.cbor.CBORGenerator;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.security.PublicKey;
import java.security.interfaces.ECPublicKey;

class CredentialData {
    public byte[] aaguid;
    public byte[] credentialId;
    public PublicKey credentialPublicKey;

    public CredentialData(byte[] aaguid, byte[] credentialId, PublicKey credentialPublicKey) {
        this.aaguid = aaguid;
        this.credentialId = credentialId;
        this.credentialPublicKey = credentialPublicKey;
    }

    public byte[] serialize() throws IOException {
        ByteArrayOutputStream output = new ByteArrayOutputStream();
        CBORGenerator gen = new CBORFactory().createGenerator(output);

        output.write(aaguid);
        output.write(credentialId);

        gen.writeStartObject(5);
        gen.writeFieldId(1);
        gen.writeNumber(2);
        gen.writeFieldId(3);
        gen.writeNumber(-7);
        gen.writeFieldId(-1);
        gen.writeNumber(1);
        gen.writeFieldId(-2);
        gen.writeBinary(((ECPublicKey) credentialPublicKey).getW().getAffineX().toByteArray());
        gen.writeFieldId(-3);
        gen.writeBinary(((ECPublicKey) credentialPublicKey).getW().getAffineY().toByteArray());
        gen.writeEndObject();

        return output.toByteArray();
    }

}

File: app/src/main/java/com/example/ctap/fido2/CredentialData.kt
package com.example.ctap.fido2

import android.util.Log
import com.example.ctap.Utils.byteArrayToHex
import com.google.common.primitives.Shorts
import java.io.IOException


/**
 * Attested credential data
 *
 * https://www.w3.org/TR/webauthn/#sec-attested-credential-data
 *
 */
class CredentialData(var aaguid: ByteArray,
                     var credentialId: ByteArray,
                     var credentialPublicKey: CredentialPublicKey) {

    private val TAG = CredentialData::class.simpleName

    @Throws(IOException::class)
    fun serialize(): ByteArray {
        val credentialIdSize = Shorts.toByteArray(credentialId.size.toShort());

        assert(aaguid.size == 16)
        assert(credentialIdSize.size == 2)
        Log.d(TAG, "credentialPublicKey=[" + byteArrayToHex(credentialPublicKey.serialize()) + "]")

        return byteArrayOf(
                *aaguid,
                *credentialIdSize,
                *credentialId,
                *credentialPublicKey.serialize()
        );
    }

}
